### PR TITLE
Refactored permission check for REST API viewsets.

### PR DESCRIPTION
**Pull request by normanjaeckel**
Please have a first look but do not merge. If this way is ok, I can implement it for every viewset we have.

This  pull request contains also some clean up.

### CODE REVIEW

**Comment by ostcar** [comment]
I line this way. Is it possible in this way, that projector_permissions is only called, when the user has not the regular permissions for the view?

Please rebase on top of master to fix travis

File: openslides/utils/rest_api.py [hunk 1]
```diff
@@ -46,8 +49,8 @@ def initial(self, request, *args, **kwargs):
 
         # Ensure that the incoming request is permitted
         self.perform_authentication(request)
-        if not self.check_projector_requirements():
-            self.check_permissions(request)
+        if not self.check_permissions(request) and not self.check_projector_requirements():
```
Comment (normanjaeckel) [inline openslides/utils/rest_api.py]: @ostcar As you see here the projector_requirements and evaluated only if the first check returns False.

**Comment by ostcar** [comment]
ok. Rebase and merge

**Comment by normanjaeckel** [comment]
Sorry, this is not ready. I will ping this when it's done.

**Comment by normanjaeckel** [comment]
@ostcar All done (with massive changes): Ready for review.

File: openslides/core/views.py [hunk 2]
```diff
@@ -265,10 +284,25 @@ def determine_metadata(self, request, view):
...
     """
     metadata_class = ConfigMetadata
 
+    def check_view_permissions(self):
+        """
+        Returns True if the user has required permissions.
+        """
+        if self.action in ('list', 'retrieve'):
+            # Everybody is able to list and retrieve the config.
+            result = True
```
Comment (ostcar) [inline openslides/core/views.py]: When anonymous user is deactivated, then this function should return False for anonymous user.

File: openslides/core/views.py [hunk 3]
```diff
@@ -168,71 +216,42 @@ def clear_elements(self, request, pk):
...
-        result = dict(openslides_version=version, plugins=[])
-        # Versions of plugins.
-        for plugin in settings.INSTALLED_PLUGINS:
-            result['plugins'].append({
-                'verbose_name': get_plugin_verbose_name(plugin),
-                'description': get_plugin_description(plugin),
-                'version': get_plugin_version(plugin)})
+        if self.action in ('list', 'retrieve'):
+            # Everybody is able to list and retrieve tags.
+            result = True
```
Comment (ostcar) [inline openslides/core/views.py]: This is False for anonymous users, when they are deactivated

File: openslides/utils/rest_api.py [hunk 4]
```diff
@@ -20,39 +19,39 @@
...
-        Runs anything that needs to occur prior to calling the method handler.
-        """
-        self.format_kwarg = self.get_format_suffix(**kwargs)
 
-        # Ensure that the incoming request is permitted
-        self.perform_authentication(request)
-        if not self.check_projector_requirements():
-            self.check_permissions(request)
-        self.check_throttles(request)
+    def get_permissions(self):
```
Comment (ostcar) [inline openslides/utils/rest_api.py]: Please add a docstring

File: openslides/utils/views.py [hunk 5]
```diff
@@ -88,14 +52,33 @@ def as_view(cls, *args, **kwargs):
...
     top_space = 3
     document_title = None
+    required_permission = None
+
+    def check_permission(self, request, *args, **kwargs):
+        """
+        Checks if the user has the required permission.
+        """
+        if self.required_permission is None:
+            return True
```
Comment (ostcar) [inline openslides/utils/views.py]: Could you add an explicit permission check. When required_permission is None, then no user should have permissions. There for we can not forget to set a permission.

File: openslides/utils/views.py [hunk 6]
```diff
@@ -88,14 +52,33 @@ def as_view(cls, *args, **kwargs):
...
+        else:
+            return request.user.has_perm(self.required_permission)
```
Comment (ostcar) [inline openslides/utils/views.py]: I like this, but I think, that you don't.

If there is a PDF that only the manager can see (for example the user Password PDF), which permission has to be defined in required_permission? Only 'can_manage' without 'can_see'?

You have either to redefine the can_manage permission to also include 'can_see', or change the required_permission attribute to an list or tuple.

I would like, that we redefine the can_manage permission. You do not have to do this in this pull request. You can leave this method as it is and create an issue to redefine the can_manage permission

File: openslides/core/views.py (hunk 2 above)
Comment (normanjaeckel) [inline openslides/core/views.py]: Done.

File: openslides/core/views.py (hunk 3 above)
Comment (normanjaeckel) [inline openslides/core/views.py]: Done.

File: openslides/utils/rest_api.py (hunk 4 above)
Comment (normanjaeckel) [inline openslides/utils/rest_api.py]: Done.

File: openslides/utils/views.py (hunk 5 above)
Comment (normanjaeckel) [inline openslides/utils/views.py]: No this is old style code for Django views. We can do this in an extra pull request.

File: openslides/utils/views.py (hunk 6 above)
Comment (normanjaeckel) [inline openslides/utils/views.py]: This is old style code (only copied from the old PermissionMixin). You can refactor this later if you like.

**Comment by normanjaeckel** [comment]
All done.

**Comment by ostcar** [comment]
ok. Rebase and merge

**Pull request merged**